keps/sig-storage/20190204-scale-testing.md

```diff
@@ -0,0 +1,170 @@
+   * [Test Cases](#test-cases)
+      * [Pod Startup](#pod-startup)
+      * [Pod Teardown](#pod-teardown)
+      * [WIP: PV Binding Tests](#wip-pv-binding-tests)
+      * [WIP: PV Provisioning Tests](#wip-pv-provisioning-tests)
+      * [WIP: PV Deletion Tests](#wip-pv-deletion-tests)
+* [Graduation Criteria](#graduation-criteria)
+   * [Phase 1](#phase-1)
+   * [Phase 2](#phase-2)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+This KEP outlines a plan for testing scalability and performance of K8s storage components.
+
+## Motivation
+
+Adding storage scale and performance tests will help:
+* Understand the current scale limits of the Kuberentes storage system.
+* Set expectations (SLOs) for consumers of the Volume API.
+* Determine bottlenecks and influence which need addressing.
+
+### Goals
+
+* Measure the overhead of K8s components owned by sig-storage:
+  * K8s volume controllers
+  * Kubelet volume manager
+  * CSI sidecars
+* Stress various dimensions of volume operations to determine:
+  * Max volumes per pod
+  * Max volumes per node
+  * Max volumes per cluster
+* Test with the following volume types:
+  * EmptyDir
+  * Secret
+  * Configmap
+  * Downward API
+  * CSI mock driver
+  * TODO: Hostpath?
+  * TODO: Local?
+* Provide tests that vendors can easily run against their volume drivers.
+
+### Non-Goals
+
+* Test and measure storage provider’s drivers.
+
+## Proposal
+
+### WIP Test Environment
+
+The tests should be developed and run in sig-scalability’s test framework and infrastructure.
+
+TODO details and links
+
+### Test Portability
+
+In order to make it easier for vendors to run these tests against their volume
+drivers, the tests for persistent volumes should only use the default
+StorageClass configured for the cluster.
+
+Therefore these tests have the following prerequisites:
+
+* A Kubernetes cluster is brought up in an appropriate environment that can
+  support the storage provider.
+* The storage backend and driver is already installed and configured.
+* A default StorageClass for that driver been installed in the cluster.
+
+### Test Cases
+
+#### Pod Startup
+
+These tests should measure how long it takes to start up a pod with volumes, assuming that the volume has already been provisioned.
+
+For each volume type:
+
+* Create many pods with 1 volume each.
+* Create 1 pod with many volumes.
+  * For PVC types, test with and without PV.NodeAffinity as one more dimension.
+  * Measure pod startup time, with a breakdown of time spent in volume scheduling, attach and mount operations.
+
+#### Pod Teardown
+
+These tests should measure how long it takes to delete a pod with volumes.
+
+For each volume type:
+
+* Delete many pods with 1 volume each.
+* Delete 1 pod with many volumes.
+  * Measure pod deletion time, with a breakdown of time spent in unmount and detach operations.
+    * Note: Detach can only be measured for CSI volumes by the removal of the VolumeAttachment object.
```

Comment:
Isn't detach always decoupled from pod deletion? I.e. pod stops, some time later the volume gets detached from the node? I remember that this was driven by the control loop in kubelet, but it's quite likely that I don't remember correctly or misunderstood.

If it is indeed decoupled from pod deletion, then an approach for measure time based on pod events won't work.

If CSI volumes need special handling, then this affects test portability.

Reply:
Yes the problem is that Pod deletion is not tied to detach.  But I think detach is still important to measure because for RWO volumes, it impacts the time it takes to launch a new Pod that uses the same volume.

We may have to make an exception on portability in order to collect detach metrics.  It could just be that for non-CSI volumes, we won't be able to collect that data.